src/chainstate/stacks/boot/exit-at-rc.clar

```diff
@@ -0,0 +1,199 @@
+(define-map voter-state
+    { address: principal }
+    {
+        proposed-rc: uint,
+        amount: uint,
+        expiration-reward-cycle: uint
+    }
+)
+
+;; What's the reward cycle number of the burnchain block height?
+;; Will runtime-abort if height is less than the first burnchain block (this is intentional).
+;; Returns uint
+(define-private (burn-height-to-reward-cycle (height uint))
+    (let
+        (
+            (local-first-burnchain-block-height (var-get first-burnchain-block-height))
+            (local-pox-reward-cycle-length (var-get pox-reward-cycle-length))
+        )
+
+        (asserts! (> height local-first-burnchain-block-height) (err ERR_BURN_BLOCK_HEIGHT_TOO_LOW))
+        (asserts! (> local-pox-reward-cycle-length u0) (err ERR_AMOUNT_NOT_POSITIVE))
+
+        (ok (/ (- height local-first-burnchain-block-height) local-pox-reward-cycle-length))
+    )
+)
+
+;; What's the current PoX reward cycle?
+;; Returns uint
+(define-private (current-pox-reward-cycle)
+    (burn-height-to-reward-cycle burn-block-height))
+
+
+;; For a specific reward cycle, this function tried to add "amount" number of votes for the proposed exit reward cycle.
+;; Note: votes are tied to reward cycles, since the vote's lifetime is tied to the voter's stacking duration.
+(define-private (add-to-rc-proposal-map (cycle-opt (optional uint)) (proposed-rc uint) (amount uint))
+    (begin
+        (match cycle-opt
+            cycle (match (map-get? rc-proposal-votes {proposed-rc: proposed-rc, curr-rc: cycle})
+                              existing-votes (map-set rc-proposal-votes {proposed-rc: proposed-rc, curr-rc: cycle} {votes: (+ amount (get votes existing-votes)) })
+
+                              ;; no existing state
+                              (map-insert rc-proposal-votes {proposed-rc: proposed-rc, curr-rc: cycle} {votes: amount})
+                          )
+            false
+        )
+    )
+)
+
+;; Used to construct a list of the reward cycles that a voter's vote would be valid for (tied to their stacking duration).
+;; Ex: if a user starts stacking at RC 18 for 3 cycles, the output of this function would be:
+;;     [ (some 18), (some 19), (some 20), none, none, none, none, none, none, none, none, none ]
+(define-private (get-voting-reward-cycles (index uint) (lock-period uint) (first-reward-cycle uint))
+    (begin
+        (if (< index lock-period)
+            (some (+ first-reward-cycle index))
+            none
+        )
+    )
+)
+
+;; TODO (#3034) - fill in real address for contract-call once known
+;; A stacking voter with no outstanding vote can call this function with their proposed exit reward cycle to vote for it.
+;; This function enforces bounds on the vote (can't be above/below specific values).
+;; If a vote is accepted, the voter can only re-vote when they stack again.
+(define-public (vote-for-exit-rc (proposed-exit-rc uint))
```

Comment:
In terms of test plans for Clarity code, one thing I've found makes code like this much more robust is to separate the validity checks into their own method, and the mutations in the success case into their own methods, and then have that validity check return all the computed state needed to carry out the mutations.  Then, you can separately and thoroughly test the validity checks without worrying about side-effects altering their behavior, and similarly, you can test the mutation code without having to mock the chainstate necessary for the validity checks to pass.

I have an example here, if you want to see how it can be done in practice: https://github.com/jcnelson/nftree/blob/main/contracts/nftree.clar#L865.  The function in question fulfills a buy order by (1) checking that the buy offer is valid (via `inner-can-fulfill-buy-offer?`), and (2) passing the result of the check into `inner-fulfill-buy-offer`.  This neatly factors testing into a side-effect-free test for the validity checks (https://github.com/jcnelson/nftree/blob/main/contracts/tests/test-nftree.clar#L1293), and a much-simpler test for the mutation (https://github.com/jcnelson/nftree/blob/main/contracts/tests/test-nftree.clar#L1380).

Reply:
Another tactic is to separate out the code from the global variables as much as possible (e.g. `block-height`, `tx-sender`, etc.) by wrapping the logic in a method that takes these values as arguments.  This makes it much easier to unit-test the logic without setting up a runtime environment for it.